### DISCUSSION TITLE
How to refresh ComboBox options when selected

### DISCUSSION THREAD

**Discussion by trampster** [AvaloniaUI/Avalonia#7592] [Q&A]
I have an App that needs to display dynamic combo box options.

I need to set those options when the user clicks the drop down, or tabs to it etc.

- I have tried GotFocus but that also occurs when the a drop down item is selected and if I update the combo box options at that point the combobox gets stuck option.
- There is a TextInputOptionsQuery which seems like the correct thing but never fires.
- The Tapped event works for clicks but not if you tab there.
- I tried databinding to the IsDropDownOpen property, but the setter never gets hit. This also wouldn't help for keyboard navigation which allows selecting an option without opening the drop down.

Currently I'm using a mixture of Tapped and GotFocus which checks for tab navigation to refresh the list. But I feed like there must be a better way which I am missing.

**Comment by timunie** [comment]
Do you have a sample you can share with us via GitHub? Maybe then your use-case will get more clear and one has a solution / idea for you.

**Reply by trampster** [reply]
It's for a serial port applications, the drop down allows the user to select from the available serial ports. However because ports can change dynamically (can be plugged in and unplugged while the application is running) Then we need to delay the population of the combo items until the user click's it / tabs to it to select a port.

**Reply by timunie** [reply]
I don't know how often this happens, but here are a few more ideas

- add a refresh button next to the control, which the user can explicitly click if the refreah should happen
- or have a background task observing the ports and update your collection if needed. 
- or add a handler in code behind to observe the isopen property, but I have to find the syntax for it

**Comment by timunie** [comment]
If you want to perform some action when the ComboBox opens you can do this: 

1. Give your ComboBox a name
2. Listen to property changes
3. do whatever you need to do

```c#
public MainWindow()
{
    ... // Other Stuff

    // We use a dispatcher, so we make sure everything is loaded before we access the control
    Dispatcher.UIThread.Post(() =>
    {
        this.Find<ComboBox>("MyComboBox").PropertyChanged += ComboBox_PropertyChanged;
    }, DispatcherPriority.Loaded);
}

private void ComboBox_PropertyChanged(object? sender, AvaloniaPropertyChangedEventArgs e)
{
    // Check for the right property
    if (e.Property == ComboBox.IsDropDownOpenProperty)
    {
        // Do your work here
        Debug.WriteLine((bool)e.NewValue ? "Opened" : "closed"); 
    }
    if (e.Property == ComboBox.IsFocusedProperty && (e.NewValue as bool?) == true)
    {
        // Or listen to IsFocusedProperty 
        Debug.WriteLine("We got the focus");
    }
}
```

Hope this helps a little bit

**Comment by mpulukkinen** [comment]
I have this exact same issue and it's really annoying that I constantly need to fight with the combobox when the list can change dynamically and everything is binded to view models
